File: assembler.py
from datetime import date
import os
import getpass
from datetime import datetime
import sys
from os.path import expanduser
import subprocess 
import getopt
import argparse
import time
import os.path
import re
from typing import NamedTuple

##000000 nop
##000001 lui R#, upper2Bytes
##000010 li  R#, lower2Bytes
##000011 add Rdest, Rsrc1, Rscrc2
##100011 sub Rdest, Rsrc1, Rscrc2
##000100 not Rdest, Rsrc1
##000101 and  Rdest, Rsrc1, Rscrc2
##000110 or Rdest, Rsrc1, Rscrc2
##000111 xor Rdest, Rsrc1, Rscrc2
##001000 xnor Rdest, Rsrc1, Rscrc2
##001001 nand Rdest, Rsrc1, Rscrc2
##001010 nor Rdest, Rsrc1, Rscrc2
##001011 shiftl Rdest, Rsrc1
##001100 shiftr Rdest, Rsrc1
##001101 Incr Rsrc
##001110 Decr Rsrc
##001111 load Rdest, [Raddress]
##010000 str Rsrc, [Raddress]
##010001 jmp @addr
##010010 jmp [R#]
##010011 jeq @addr
##010100 jeq [R#]
##010101 jneq @addr
##010110 jneq [R#]
##010111 jg @addr
##011000 jg [R#]
##011001 jge @addr
##011010 jgw [R#]
##011011 jl @addr
##011100 jl [R#]
##011101 jle @addr
##011110 jle [R#]
##011111 cmp
##100010 mov

opcode = {'nop'    :'000000',
          'lui'    :'000001',
          'li'     :'000010',
          'add'    :'000011',
          'sub'    :'100011', 
          'not'    :'000100', 
          'and'    :'000101',
          'or'     :'000110', 
          'xor'    :'000111', 
          'xnor'   :'001000',
          'nand'   :'001001',
          'nor'    :'001010',
          'shiftl' :'001011',
          'shiftr' :'001100',
          'Incr'   :'001101',
          'Decr'   :'001110',
          'load'   :'001111',
          'str'    :'010000',
          'jmp'    :'010001',
          'jeq'    :'010011',
          'jneq'   :'010101',
          'jg'     :'010111',
          'jge'    :'011001',
          'jl'     :'011011',
          'jle'    :'011101',
          'cmp'    :'011111',
          'call'   :'100000',
          'return' :'100001',  
          'mov'    :'100010',
          'jmpa'   :'010100'
          }
           
registers = {'R0':"00000", 
             'R1':"00001",
             'R2':'00010',
             'R3':'00011',
             'R4':'00100',
             'R5':'00101',
             'R6':'00110',
             'R7':'00111',
             'R8':'01000',
             'R9':'01001',
             'R10':'01010',
             'R11':"01011", 
             'R12':"01100",
             'R13':'01101',
             'R14':'01110',
             'R15':'01111',
             'R16':'10000',
             'R17':'10001',
             'R18':'10010',
             'R19':'10011',
             'R20':'10100',
             'R21':'10101',
             'R22':'10110',
             'R23':'10111',
             'R24':'11000',
             'R25':'11001',
             'R26':'11010',
             'R27':'11011',
             'R28':'11100',
             'R29':'11101',
             'R30':'11110',
             'R31':'11111'}

symbol_table = {}
program = []

def pass1(assemblyFile) :
    lines = []
    identifierAddress = 16
    with open(assemblyFile) as f:
        lines = f.readlines()
    pc = 0
    for line in lines:
        line = line.lstrip()
        line = line.replace('\n', '')
       
        if (line.startswith("#") == False and line !="" and line.startswith("//") == False):
            
            components = re.split(',?;?\s+', line)
            print(components)


            if (line.startswith('(') == True):
                labelName = line.replace('(','')
                labelName = labelName.replace(')','')
                symbol_table[labelName] = pc
            elif(components[0] == "li" and components[2].isnumeric() == False):
                symbol_table[components[2]] = identifierAddress
                identifierAddress += 1
                pc = pc + 2
            elif(components[0] == "li" and components[2].isnumeric() == True):
                pc = pc + 2
            else:
                pc = pc + 1
    print(symbol_table)
    print("number of instructions = " + str(pc))

def generate_memory_list():
    f = open("memory.list", "w")
    for assemCode in program:
        if (assemCode.startswith("--")== False):
            decimal = int(assemCode, 2)
            hex_string = hex(decimal)
            hex_string = hex_string.replace('0x', '')
            hex_string = hex_string.zfill(8)
            print(hex_string) 
            f.write(hex_string + '\n')


def generate_rom_vhdl():
    f = open("rom.vhd", "w")
    f.write("-------------------------------------------\n")
    f.write("---- ROM example\n")
    f.write("-------------------------------------------\n")
    f.write("library ieee;\n")
    f.write("use ieee.std_logic_1164.all;\n")
    f.write("use ieee.numeric_std.all;\n")
    f.write("-------------------------------------------\n")
    f.write("entity rom is\n")
    f.write("   port(\n")
    f.write("         address: in std_logic_vector(31 downto 0);--integer range 0 to 31;\n")
    f.write("         rom_out: out std_logic_vector(31 downto 0)\n")
    f.write("       );\n")
    f.write("end entity;\n")
    f.write("--------------------------------------------\n")
    f.write("architecture rtl of rom is\n")
    f.write("   signal address_reg : std_logic_vector(31 downto 0);\n")
    f.write("   type memory is array (0 to 1024) of std_logic_vector(31 downto 0);\n")
    f.write("   \n")
    f.write("   constant myrom :memory :=\n")
    f.write("   (\n")
    pc = 0
    for assemCode in program:
        if (assemCode.startswith("--")):
            f.write('      '  + assemCode)
        else:
            f.write('      '  + str(pc) + '=>"' + assemCode + '",\n')
            pc += 1
    f.write("\n")
    f.write("      others=>\"00000000000000000000000000000000\"\n")
    f.write("   );\n")
    f.write("   \n")
    f.write("begin\n")
    f.write("   process (address)\n")
    f.write("   begin\n")
    f.write("         rom_out <= myrom(to_integer(unsigned(address)));\n")
    f.write("   end process;\n")
    f.write("   \n")
    f.write("end rtl;\n")
    

def assemble(assemblyFile):
    lines = []
    with open(assemblyFile) as f:
        lines = f.readlines()
    f = open("assembly.txt", "w")
       
    for line in lines:
        line = line.replace('\n', '')
        line = line.lstrip()
        
        if (line.startswith("#") == False and line !="" and line.startswith("//") == False and line.startswith("(") == False):
            components = re.split(',?;?\s+', line)
            for elem in components:
               if(elem ==''):
                  components.remove('')
            #no operation
            if (components[0] == "nop"):
                program.append("--" + line + "\n")
                program.append(components["return"] + "00000000000000000000000000")

            if (components[0] == "call"):
                if (components[1].isnumeric()):
                    hexBinary = bin(int(components[1], 10))[2:].zfill(26)
                    program.append("--" + line + "\n")
                    program.append(opcode["call"] + hexBinary)
                else:
                    address = str(symbol_table[components[1]])
                    hexBinary = bin(int(address, 10))[2:].zfill(26)
                    program.append("--" + line + "\n")
                    program.append(opcode["call"] + hexBinary)

            if (components[0] == "return"):
                program.append("--" + line + "\n")
                program.append(opcode["return"] + "00000000000000000000000000")

            #lui or li operation
            elif (components[0] == "li"):
                if (components[2].startswith("_") == False):
                    literalVal = components[2].replace('0x','')
                 
                    literalVal = bin(int(literalVal))
                    literalVal = literalVal.replace('b','')
                    upperHex = literalVal[16:31].zfill(21)
                    lowerHex = literalVal[0:15].zfill(21)
                    program.append("--" + line + "\n")
                    program.append(opcode["lui"] + registers[components[1]] + str(upperHex))
                    program.append(opcode["li"] + registers[components[1]] + str(lowerHex))
                else:
                    literalVal = str(symbol_table[components[2]])
                    #if (literalVal.isnumeric()):
                    literalVal = literalVal.zfill(8)
                    upperHex = literalVal[0:4]
                    lowerHex = literalVal[4:9] 
                    upperHexBinary = bin(int(upperHex, 10))[2:].zfill(21)
                    lowerHexBinary = bin(int(lowerHex, 10))[2:].zfill(21)
                    program.append("--" + line + "\n")
                    program.append(opcode["lui"] + registers[components[1]] + upperHexBinary)
                    program.append(opcode["li"] + registers[components[1]] + lowerHexBinary)


            if(components[0] == "jmp" or components[0] == "jeq" or components[0] == "jneq" or components[0] == "jg" or components[0] == "jge" or components[0] == "jl" or components[0] == "jle"):
                if (components[1].isnumeric()):
                    hexBinary = bin(int(components[1], 10))[2:].zfill(26)
                    program.append("--" + line + "\n")
                    program.append(opcode[components[0]] + hexBinary)
                else:
                    address = str(symbol_table[components[1]])
                    hexBinary = bin(int(address, 10))[2:].zfill(26)
                    program.append("--" + line + "\n")
                    program.append(opcode[components[0]] + hexBinary)

            if(components[0] == "add" or components[0] == "sub" or components[0] == "and" or components[0] == "or" or components[0] == "xor" or components[0] == "xnor" or components[0] == "nand" or components[0] == "nor"):
                program.append("--" + line + "\n")
                program.append(opcode[components[0]] + registers[components[1]] + registers[components[2]] +  registers[components[3]] +'00000000000')

            if(components[0] == "shiftl" or components[0] == "shiftr"):
                program.append("--" + line + "\n")
                program.append(opcode[components[0]] + registers[components[1]] + registers[components[2]] + "0000000000000000")
            
            if(components[0] == "Incr" or components[0] == "Decr"):
                program.append("--" + line + "\n")
                program.append(opcode[components[0]] + registers[components[1]] + "000000000000000000000")

            if(components[0] == "not"):
                 program.append("--" + line + "\n")
                 program.append(opcode[components[0]] + registers[components[1]] + registers[components[2]] + "0000000000000000")

            if(components[0] == "load" or components[0] == "str" or components[0] == "cmp"):
                program.append("--" + line + "\n")
                program.append(opcode[components[0]] + registers[components[1]] + registers[components[2]] + "0000000000000000")

            if(components[0] == "mov"):
                program.append("--" + line + "\n")
                program.append(opcode[components[0]] + registers[components[1]] + registers[components[2]] + "0000000000000000")

            if(components[0] == "jmpa"):
                 program.append("--" + line + "\n")
                 program.append(opcode[components[0]] + registers[components[1]] + "000000000000000000000")


    #generate_rom_vhdl()
    generate_memory_list()
    f.close()


def main():
    arguments = len(sys.argv)
    pass1("assembly.asm")
    assemble("assembly.asm")


if __name__ == "__main__":
    main() 

File: SimpleX_Compiler/assembler.cpp
#include "assembler.hpp"
#include <fstream>
#include <regex>
#include <iostream>
#include <algorithm>

Assembler::Assembler()
{
    opcode_["nop"]    = "000000";
    opcode_["lui"]    = "000001";
    opcode_["li"]     = "000010";
    opcode_["add"]    = "000011";
    opcode_["sub"]    = "100011"; 
    opcode_["not"]    = "000100"; 
    opcode_["and"]    = "000101";
    opcode_["or"]     = "000110"; 
    opcode_["xor"]    = "000111"; 
    opcode_["xnor"]   = "001000";
    opcode_["nand"]   = "001001";
    opcode_["nor"]    = "001010";
    opcode_["shiftl"] = "001011";
    opcode_["shiftr"] = "001100";
    opcode_["Incr"]   = "001101";
    opcode_["Decr"]   = "001110";
    opcode_["load"]   = "001111";
    opcode_["str"]    = "010000";
    opcode_["jmp"]    = "010001";
    opcode_["jeq"]    = "010011";
    opcode_["jneq"]   = "010101";
    opcode_["jg"]     = "010111";
    opcode_["jge"]    = "011001";
    opcode_["jl"]     = "011011";
    opcode_["jle"]    = "011101";
    opcode_["cmp"]    = "011111";
    opcode_["call"]   = "100000";
    opcode_["return"] = "100001";  
    opcode_["mov"]    = "100010";
    opcode_["jmpa"]   = "010100";
    opcode_["mul"]    = "100100";
           
    registers_["R0"] = "00000"; 
    registers_["R1"] = "00001";
    registers_["R2"] = "00010";
    registers_["R3"] = "00011";
    registers_["R4"] = "00100";
    registers_["R5"] = "00101";
    registers_["R6"] = "00110";
    registers_["R7"] = "00111";
    registers_["R8"] = "01000";
    registers_["R9"] = "01001";
    registers_["R10"] = "01010";
    registers_["R11"] = "01011"; 
    registers_["R12"] = "01100";
    registers_["R13"] = "01101";
    registers_["R14"] = "01110";
    registers_["R15"] = "01111";
    registers_["R16"] = "10000";
    registers_["R17"] = "10001";
    registers_["R18"] = "10010";
    registers_["R19"] = "10011";
    registers_["R20"] = "10100";
    registers_["R21"] = "10101";
    registers_["R22"] = "10110";
    registers_["R23"] = "10111";
    registers_["R24"] = "11000";
    registers_["R25"] = "11001";
    registers_["R26"] = "11010";
    registers_["R27"] = "11011";
    registers_["R28"] = "11100";
    registers_["R29"] = "11101";
    registers_["R30"] = "11110";
    registers_["R31"] = "11111";
}

Assembler::~Assembler()
{}


std::vector<std::string> Assembler::splitString(std::string text)
{
   std::vector<std::string> words{};
   size_t pos = 0;
   std::string space_delimiter = " ";
   while ((pos = text.find(space_delimiter)) != std::string::npos || (pos = text.find(',')) != std::string::npos)
   {
    if (text.substr(0, pos) != "")
    {
       std::replace(text.begin(), text.end(), ',', ' '); 
       words.push_back(text.substr(0, pos));
    }
    text.erase(0, pos + space_delimiter.length());
   }
   words.push_back(text);

   return words;
}

bool is_number(const std::string& s)
{
    std::string::const_iterator it = s.begin();
    while (it != s.end() && std::isdigit(*it)) ++it;
    return !s.empty() && it == s.end();
}

const std::string WHITESPACE = " \n\r\t\f\v";

std::string ltrim(const std::string &s)
{
    size_t start = s.find_first_not_of(WHITESPACE);
    return (start == std::string::npos) ? "" : s.substr(start);
}
 
std::string rtrim(const std::string &s)
{
    size_t end = s.find_last_not_of(WHITESPACE);
    return (end == std::string::npos) ? "" : s.substr(0, end + 1);
}

std::string trim(const std::string &s) 
{
    return rtrim(ltrim(s));
}


void Assembler::pass1(std::string assemblyFile) 
{
   uint32_t pc = 0;

   std::ifstream infile(assemblyFile);
   std::string line;
   asm_ = std::ofstream();
   asm_.open("memory.list");
  
   while(std::getline(infile, line))
   {
      line = trim(line);
      if (line != "" && line.rfind("/", 0) != 0)
      {
         std::vector<std::string> components = splitString(line);
         if (line.rfind("(", 0) == 0)
         {
             line.erase(remove(line.begin(), line.end(), '('), line.end()); 
             line.erase(remove(line.begin(), line.end(), ')'), line.end()); 
             labelTable_[line] = pc;
         }
         else if (components[0] == "li" && is_number(components[2]))
         {
             pc = pc + 2;
         }
         else if (components[0] == "li" && (is_number(components[2]) == false))
         {
             pc = pc + 2;
         }
         else
         {
             pc = pc + 1;
         }
      }
   }
}

std::string stringToBinary(std::string decimal, uint8_t numOfbits) 
{
    std::string binaryString;
    uint32_t decimalInt = atoi(decimal.c_str());
    for (int i = 0; i < numOfbits; i++)
    {
         if ((decimalInt % 2) == 0)
         {
            binaryString.push_back('0');
         }
         else
         {
            binaryString.push_back('1');
         }
         decimalInt = decimalInt / 2;
    }

    std::reverse(binaryString.begin(), binaryString.end()); 

    return binaryString;
}

void Assembler::CreateHexFile()
{
    hexProgram_.clear();
    for(size_t i = 0; i < program_.size() ; i++)
    {
        std::string binToHex, tmp = "0000";
        for (size_t j = 0; j < program_.at(i).size(); j += 4)
        {
            if (program_.at(i).rfind("-", 0) != 0)
            {
                tmp = program_.at(i).substr(j, 4);
                if  (!tmp.compare("0000")) 
                    binToHex += "0";
                else if (!tmp.compare("0001"))
                    binToHex += "1";
                else if (!tmp.compare("0010")) 
                    binToHex += "2";
                else if (!tmp.compare("0011")) 
                    binToHex += "3";
                else if (!tmp.compare("0100"))
                     binToHex += "4";
                else if (!tmp.compare("0101")) 
                    binToHex += "5";
                else if (!tmp.compare("0110")) 
                    binToHex += "6";
                else if (!tmp.compare("0111")) 
                    binToHex += "7";
                else if (!tmp.compare("1000")) 
                    binToHex += "8";
                else if (!tmp.compare("1001")) 
                    binToHex += "9";
                else if (!tmp.compare("1010")) 
                    binToHex += "a";
                else if (!tmp.compare("1011")) 
                    binToHex += "b";
                else if (!tmp.compare("1100")) 
                    binToHex += "c";
                else if (!tmp.compare("1101")) 
                    binToHex += "d";
                else if (!tmp.compare("1110")) 
                    binToHex += "e";
                else if (!tmp.compare("1111")) 
                    binToHex += "f";
                else continue;
            }
        }
        if (binToHex != "")
        {
            hexProgram_.push_back(binToHex);
        }
    }

    std::ofstream fileOut;
    remove("memory.list");
    fileOut.open("memory.list");
    for (int i = 0; i < hexProgram_.size(); i++)
    {
       fileOut << hexProgram_.at(i) << std::endl;
    }
}


void Assembler::assemble(std::string assemblyFile)
{
   std::ifstream infile(assemblyFile);
   std::string line;
   
   while(std::getline(infile, line))
   {
      line = trim(line);
      if (line.rfind("#", 0) != 0 && line != "" && line.rfind("/", 0) != 0 && line.rfind("(", 0) != 0 )
      {
           std::vector<std::string> components = splitString(line);
        
            //no operation
            if (components[0] == "nop")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_["nop"] + "00000000000000000000000000");
            }
            if (components[0] == "call")
            {
                if (is_number(components[1]))
                {
                    std::string hexBinary = stringToBinary(components[1], 26);
                    program_.push_back("--" + line);
                    program_.push_back(opcode_["call"] + hexBinary);
                }
                else
                {
                    uint32_t addressInt = (labelTable_[components[1]]);
                    std::string address = std::to_string(addressInt);
                    std::string hexBinary = stringToBinary(address, 26);//bin(int(address, 10))[2:].zfill(26)
                    program_.push_back("--" + line);
                    program_.push_back(opcode_["call"] + hexBinary);
                }
            }
            if (components[0] == "return")
            {
                program_.push_back("--" + line );
                program_.push_back(opcode_["return"] + "00000000000000000000000000");
            }
            //lui or li operation
            else if (components[0] == "li")
            {
                if (components[2].rfind("_", 0) != 0)
                {
                    std::string literalVal = components[2];
                    literalVal = stringToBinary(literalVal, 32);
                    std::string upperHex = "00000" + literalVal.substr(0, 16);
                    std::string lowerHex = "00000" + literalVal.substr(16, 16);
                    program_.push_back("--" + line);
                    program_.push_back(opcode_["lui"] + registers_[components[1]] + upperHex);
                    program_.push_back(opcode_["li"] + registers_[components[1]] + lowerHex);
                }
                else
                {
                    uint32_t literalValInt = labelTable_[components[2]];
                    std::string literalVal = std::to_string(literalValInt);
                    literalVal = stringToBinary(literalVal, 32);
                    std::string upperHex = "00000" + literalVal.substr(0, 16);//[0:4]
                    std::string lowerHex = "00000" + literalVal.substr(16, 16);//[4:9] 
    
                    program_.push_back("--" + line);
                    program_.push_back(opcode_["lui"] + registers_[components[1]] + upperHex);
                    program_.push_back(opcode_["li"] + registers_[components[1]] + lowerHex);
                }
            }
            if(components[0] == "jmp" or components[0] == "jeq" or components[0] == "jneq" or components[0] == "jg" or components[0] == "jge" or components[0] == "jl" or components[0] == "jle")
            {
                if (is_number(components[1]))
                {
                    std::string hexBinary = stringToBinary(components[1], 26);//bin(int(components[1], 10))[2:].zfill(26)
                    program_.push_back("--" + line);
                    program_.push_back(opcode_[components[0]] + hexBinary);
                }
                else
                {
                    uint32_t literalValInt = labelTable_[components[1]];
                    std::string address = std::to_string(literalValInt);
                    std::string hexBinary = stringToBinary(address, 26);//[2:].zfill(26)
                    program_.push_back("--" + line);
                    program_.push_back(opcode_[components[0]] + hexBinary);
                }
            }
            if(components[0] == "add" or components[0] == "sub" or components[0] == "and" or components[0] == "or" or components[0] == "xor" or components[0] == "xnor" or components[0] == "nand" or components[0] == "nor" or components[0] == "mul")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_[components[0]] + registers_[components[1]] + registers_[components[2]] +  registers_[components[3]] + "00000000000");
            }
            if(components[0] == "shiftl" or components[0] == "shiftr")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_[components[0]] + registers_[components[1]] + registers_[components[2]] + "0000000000000000");
            }
            if(components[0] == "Incr" or components[0] == "Decr")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_[components[0]] + registers_[components[1]] + "000000000000000000000");
            }
            if(components[0] == "not")
            {
                 program_.push_back("--" + line);
                 program_.push_back(opcode_[components[0]] + registers_[components[1]] + registers_[components[2]] + "0000000000000000");
            }
            if(components[0] == "load" or components[0] == "str" or components[0] == "cmp")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_[components[0]] + registers_[components[1]] + registers_[components[2]] + "0000000000000000");
            }
            if(components[0] == "mov")
            {
                program_.push_back("--" + line);
                program_.push_back(opcode_[components[0]] + registers_[components[1]] + registers_[components[2]] + "0000000000000000");
            }
            if(components[0] == "jmpa")
            {
                 program_.push_back("--" + line);
                 program_.push_back(opcode_[components[0]] + registers_[components[1]] + "000000000000000000000");
            }
      }
   }
   infile.close();

   CreateHexFile();
}
